File: src/test_core/fmg/common/notifier/PropertyChangeExecutorTest.java
package fmg.common.notifier;

import java.beans.PropertyChangeListener;

import org.junit.*;

import fmg.common.Logger;
import fmg.core.mosaic.MosaicModelTest;
import io.reactivex.Flowable;

public class PropertyChangeExecutorTest {

    private static class SimpleDataObj implements INotifyPropertyChanged, AutoCloseable {

        private final NotifyPropertyChanged notifier = new NotifyPropertyChanged(this);

        @Override
        public void addListener(PropertyChangeListener listener) { notifier.addListener(listener); }
        @Override
        public void removeListener(PropertyChangeListener listener) { notifier.removeListener(listener); }

        @Override
        public void close() { notifier.close(); }

        public boolean isDisposed() { return notifier.isDisposed(); }

    }


    @BeforeClass
    public static void setup() {
        Logger.info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        Logger.info("> PropertyChangeExecutorTest::setup");

        MosaicModelTest.ProjSettings();

        Flowable.just("UI factory inited...").subscribe(Logger::info);
    }

    @Before
    public void before() {
        Logger.info("======================================================");
    }

    @AfterClass
    public static void after() {
        Logger.info("======================================================");
        Logger.info("< PropertyChangeExecutorTest closed");
        Logger.info("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }

    @Test
    public void simpleUsageTest() {
        Logger.info("> PropertyChangeExecutorTest::simpleUsageTest");

        SimpleDataObj[] d = { null };
        new PropertyChangeExecutor<>(() -> d[0] = new SimpleDataObj()).run(10, 1000,
            data -> {
                Logger.info("    data modificator");
            }, (data, modifiedProperties) -> {
                Logger.info("    data validator");
                Assert.assertNotNull(data);
                Assert.assertEquals(0, modifiedProperties.size());
                Assert.assertFalse(data.isDisposed());
            });
        Assert.assertNotNull(d[0]);
        Assert.assertTrue(d[0].isDisposed());
    }

    @Test
    public void extendedUsageTest() {
        Logger.info("> PropertyChangeExecutorTest::extendedUsageTest");

        SimpleDataObj[] d = { null };
        new PropertyChangeExecutor<>(() -> d[0] = new SimpleDataObj(), false).run(300, 1000,
            data -> {
                Logger.info("    data modificator");
            }, (data, modifiedProperties) -> {
                Logger.info("    data validator");
                Assert.assertNotNull(data);
                Assert.assertEquals(0, modifiedProperties.size());
                Assert.assertFalse(data.isDisposed());
            });
        Assert.assertNotNull(d[0]);
        Assert.assertFalse(d[0].isDisposed());

        new PropertyChangeExecutor<>(() -> d[0]).run(10, 1000,
            data -> {
                Logger.info("    data modificator");
            }, (data, modifiedProperties) -> {
                Logger.info("    data validator");
                Assert.assertEquals(0, modifiedProperties.size());
                Assert.assertFalse(data.isDisposed());
            });
        Assert.assertTrue(d[0].isDisposed());
    }

    @Test
    public void creatorFailTest() {
        Logger.info("> PropertyChangeExecutorTest::creatorFailTest");

        IllegalArgumentException failEx = new IllegalArgumentException("Tested exception");
        try {
            new PropertyChangeExecutor<>(() -> { throw failEx; }).run(10, 1000,
               data -> {
                   Logger.info("    data modificator");
                   Assert.fail();
               }, (data, modifiedProperties) -> {
                   Logger.info("    data validator");
                   Assert.fail();
               });
            Assert.fail();
        } catch(Throwable ex) {
            Assert.assertEquals(failEx, ex);
        }
    }

    @Test
    public void modificatorFailTest() {
        Logger.info("> PropertyChangeExecutorTest::modificatorFailTest");

        IllegalArgumentException failEx = new IllegalArgumentException("Tested exception");
        try {
            new PropertyChangeExecutor<>(SimpleDataObj::new).run(10, 1000,
               data -> {
                   Logger.info("    data modificator");
                   throw failEx;
               }, (data, modifiedProperties) -> {
                   Logger.info("    data validator");
                   Assert.fail();
               });
            Assert.fail();
        } catch(Throwable ex) {
            Assert.assertEquals(failEx, ex);
        }
    }


    @Test
    public void validatorFailTest() {
        Logger.info("> PropertyChangeExecutorTest::modificatorFailTest");

        IllegalArgumentException failEx = new IllegalArgumentException("Tested exception");
        try {
            new PropertyChangeExecutor<>(SimpleDataObj::new).run(10, 1000,
               data -> {
                   Logger.info("    data modificator");
               }, (data, modifiedProperties) -> {
                   Logger.info("    data validator");
                   throw failEx;
               });
            Assert.fail();
        } catch(Throwable ex) {
            Assert.assertEquals(failEx, ex);
        }
    }

}

File: src/test_core/fmg/common/notifier/PropertyChangeExecutorTest.cs
using System;
using System.ComponentModel;
using System.Threading.Tasks;

namespace Fmg.Common.Notifier {

    public abstract class PropertyChangeExecutorTest {

        private class SimpleDataObj : INotifyPropertyChanged, IDisposable {

            public event PropertyChangedEventHandler PropertyChanged {
                add    { _notifier.PropertyChanged += value;  }
                remove { _notifier.PropertyChanged -= value;  }
            }
            private readonly NotifyPropertyChanged _notifier;

            internal SimpleDataObj() {
                _notifier = new NotifyPropertyChanged(this);
            }

            public void Dispose() {
                _notifier.Dispose();
            }

            public bool Disposed => _notifier.Disposed;

        }

        protected abstract void AssertEqual(int expected, int actual);
        protected abstract void AssertNotNull(object anObject);
        protected abstract void AssertTrue(bool condition);
        protected abstract void AssertFalse(bool condition);
        protected abstract void AssertFail();

        public virtual void Setup() {
            Logger.Info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            Logger.Info(">" + nameof(PropertyChangeExecutorTest) + "::" + nameof(Setup));
        }

        public virtual void Before() {
            Logger.Info("======================================================");
        }

        public virtual void After() {
            Logger.Info("======================================================");
            Logger.Info("< " + nameof(PropertyChangeExecutorTest) + " closed");
            Logger.Info("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        public virtual async Task SimpleUsageTest() {
            Logger.Info("> PropertyChangeExecutorTest::SimpleUsageTest");

            SimpleDataObj d = null;
            await new PropertyChangeExecutor<SimpleDataObj>(() => d = new SimpleDataObj()).Run(10, 1000,
                data => {
                    Logger.Info("    data modificator");
                }, (data, modifiedProperties) => {
                    Logger.Info("    data validator");
                    AssertNotNull(data);
                    AssertEqual(0, modifiedProperties.Count);
                    AssertFalse(data.Disposed);
                });
            AssertNotNull(d);
            AssertTrue(d.Disposed);
        }

        public virtual async Task ExtendedUsageTest() {
            Logger.Info("> PropertyChangeExecutorTest::ExtendedUsageTest");

            SimpleDataObj d = null;
            await new PropertyChangeExecutor<SimpleDataObj>(() => d = new SimpleDataObj(), false).Run(10, 1000,
                data => {
                    Logger.Info("    data modificator");
                }, (data, modifiedProperties) => {
                    Logger.Info("    data validator");
                    AssertNotNull(data);
                    AssertEqual(0, modifiedProperties.Count);
                    AssertFalse(data.Disposed);
                });
            AssertNotNull(d);
            AssertFalse(d.Disposed);

            await new PropertyChangeExecutor<SimpleDataObj>(() => d).Run(10, 1000,
                data => {
                    Logger.Info("    data modificator");
                }, (data, modifiedProperties) => {
                    Logger.Info("    data validator");
                    AssertEqual(0, modifiedProperties.Count);
                    AssertFalse(data.Disposed);
                });
            AssertTrue(d.Disposed);
        }

        public virtual async Task CreatorFailTest() {
            Logger.Info("> PropertyChangeExecutorTest::CreatorFailTest");

            var failEx = new ArgumentException("Tested exception");
            try {
                await new PropertyChangeExecutor<SimpleDataObj>(() => { throw failEx; }).Run(10, 1000,
                   data => {
                       Logger.Info("    data modificator");
                       AssertFail();
                   }, (data, modifiedProperties) => {
                       Logger.Info("    data validator");
                       AssertFail();
                   });
                AssertFail();
            } catch(Exception ex) {
                //AssertEqual(failEx, ex);
                AssertTrue(ReferenceEquals(failEx, ex));
            }
        }

        public virtual async Task ModificatorFailTest() {
            Logger.Info("> PropertyChangeExecutorTest::ModificatorFailTest");

            var failEx = new ArgumentException("Tested exception");
            try {
                await new PropertyChangeExecutor<SimpleDataObj>(() => new SimpleDataObj()).Run(10, 1000,
                   data => {
                       Logger.Info("    data modificator");
                       throw failEx;
                   }, (data, modifiedProperties) => {
                       Logger.Info("    data validator");
                       AssertFail();
                   });
                AssertFail();
            } catch(Exception ex) {
                //AssertEqual(failEx, ex);
                AssertTrue(ReferenceEquals(failEx, ex));
            }
        }

        public virtual async Task ValidatorFailTest() {
            Logger.Info("> PropertyChangeExecutorTest::ModificatorFailTest");

            var failEx = new ArgumentException("Tested exception");
            try {
                await new PropertyChangeExecutor<SimpleDataObj>(() => new SimpleDataObj()).Run(10, 1000,
                   data => {
                       Logger.Info("    data modificator");
                   }, (data, modifiedProperties) => {
                       Logger.Info("    data validator");
                       throw failEx;
                   });
                AssertFail();
            } catch(Exception ex) {
                //AssertEqual(failEx, ex);
                AssertTrue(ReferenceEquals(failEx, ex));
            }
        }

    }

}
